core/trino-main/src/main/java/io/trino/server/protocol/Query.java

```diff
@@ -343,6 +343,14 @@ public synchronized ListenableFuture<QueryResults> waitForResults(long token, Ur
         return Futures.transform(futureStateChange, ignored -> getNextResult(token, uriInfo, targetResultSize), resultsProcessorExecutor);
     }
 
+    public synchronized void markResultsConsumedIfReady()
+    {
+        if (!resultsConsumed && exchangeDataSource.isFinished()) {
+            queryManager.resultsConsumed(queryId);
+            resultsConsumed = true;
```

Comment:
I'm afraid that might break the logic. I wonder if this boolean has to be set here or is it sufficient to call `queryManager.resultsConsumed` to let the query "finish"?